core/src/main/java/io/cucumber/core/runtime/Runtime.java

```diff
@@ -87,22 +93,14 @@ public void run() {
             bus.send(new TestSourceRead(bus.getInstant(), feature.getUri().toString(), feature.getSource()));
         }
 
-        final List<PickleEvent> filteredEvents = new ArrayList<>();
-        for (CucumberFeature feature : features) {
-            for (final PickleEvent pickleEvent : feature.getPickles()) {
-                if (filters.matchesFilters(pickleEvent)) {
-                    filteredEvents.add(pickleEvent);
-                }
-            }
-        }
-
-        final List<PickleEvent> orderedEvents = pickleOrder.orderPickleEvents(filteredEvents);
-        final List<PickleEvent> limitedEvents = filters.limitPickleEvents(orderedEvents);
-
-        final List<Future<?>> executingPickles = new ArrayList<>();
-        for (final PickleEvent pickleEvent : limitedEvents) {
-            executingPickles.add(executor.submit(() -> runnerSupplier.get().runPickle(pickleEvent)));
-        }
+        final List<Future<?>> executingPickles = features.stream()
+                .flatMap(feature -> feature.getPickles().stream())
+                .filter(filter)
+                .collect(Collectors.collectingAndThen(Collectors.toList(),
+                        list -> pickleOrder.orderPickleEvents(list).stream()))
+                .limit(limit > 0 ? limit : Integer.MAX_VALUE)
+                .map(pickleEvent -> executor.submit(() -> runnerSupplier.get().runPickle(pickleEvent)))
+                .collect(Collectors.toList());
```

Comment:
`collectingAndThen` neat!

Reply:
Yes that one was new to me as well! Needed it since the random sort order with seed couldn't use a `Comparator`, so learned a new trick. :)

Reply:
Using comparator wouldn't have worked anyway. Java will complain if the comparator isn't consistent.